### PR TITLE
Prototype using const generic for simd_shuffle IDX array

### PR DESCRIPTION
**Pull request by oli-obk**
cc https://github.com/rust-lang/rust/issues/85229

r? @workingjubilee on the design

TLDR: there is now a `fn simd_shuffle_generic<T, U, const IDX: &'static [u32]>(x: T, y: T) -> U;` intrinsic that allows replacing

```rust
simd_shuffle(a, b, const { stuff })
```

with

```rust
simd_shuffle_generic::<_, _, {&stuff}>(a, b)
```

which makes the compiler implementations much simpler, if we manage to at some point eliminate `simd_shuffle`.

There are some issues with this today though (can't do math without bubbling it up in the generic arguments). With this change, we can start porting the simple cases and get better data on the others.

### CODE REVIEW

**Comment by RalfJung** [comment]
Miri code LGTM. Would be nice to have a test case actually calling the intrinsic (in tests/pass/portable-simd.rs).

**Comment by workingjubilee** [comment]
Should this wait on that test case?

File: tests/ui/simd/intrinsic/generic-elements.rs [hunk 1]
```diff
@@ -71,5 +73,29 @@ fn main() {
...
+//~^ ERROR element type `i32` (element of input `i32x4`), found `f32x4` with element type `f32`
+        simd_shuffle_generic::<_, f32x8, I8>(x, x);
+//~^ ERROR element type `i32` (element of input `i32x4`), found `f32x8` with element type `f32`
+
+        simd_shuffle_generic::<_, i32x8, I2>(x, x);
+        //~^ ERROR expected return type of length 2, found `i32x8` with length 8
+        simd_shuffle_generic::<_, i32x8, I4>(x, x);
+        //~^ ERROR expected return type of length 4, found `i32x8` with length 8
+        simd_shuffle_generic::<_, i32x2, I8>(x, x);
+        //~^ ERROR expected return type of length 8, found `i32x2` with length 2
```
Comment (workingjubilee) [inline tests/ui/simd/intrinsic/generic-elements.rs:99]: This seems wrong, it shouldn't error? Given:
- lhs TxN vector
- rhs TxN vector
- i1xM mask vector

the shufflevector instruction produces a vector of type T but size **M**, and M > N is legitimate (especially in the case of "unifying two vectors where `M == N*2`").

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 2]
```diff
@@ -1030,6 +1033,56 @@ fn generic_simd_intrinsic<'ll, 'tcx>(
...
+    if name == sym::simd_shuffle_generic {
+        let idx = fn_args[2]
+            .expect_const()
+            .eval(tcx, ty::ParamEnv::reveal_all(), Some(span))
+            .unwrap()
+            .unwrap_branch();
+        let n = idx.len() as u64;
+
+        require_simd!(ret_ty, InvalidMonomorphization::SimdReturn { span, name, ty: ret_ty });
+        let (out_len, out_ty) = ret_ty.simd_size_and_type(bx.tcx());
```
Comment (RalfJung) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:1045]: `(ret_len, ret_elem)` would be more consistent with the in/out variables, I think?

File: tests/ui/simd/intrinsic/generic-elements.rs (hunk 1 above)
Comment (RalfJung) [reply tests/ui/simd/intrinsic/generic-elements.rs:99]: `I8` has length 8, the return type has length 2, so `M` here would be both 2 and 8 -- and hence it should error, no?

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:1045]: it's copy paste from the `simd_shuffle` impl, but yes

File: tests/ui/simd/intrinsic/generic-elements.rs (hunk 1 above)
Comment (workingjubilee) [reply tests/ui/simd/intrinsic/generic-elements.rs:99]: oh, I see. apologies, yes, that's correct. Hmm that's... annoying, but anything more fancy is probably too painful to implement at this level of the type system.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:1045]: There are 8 occurences now (there were 7) of this same code. We should probably deduplicate something here and reconsider the names, but not in this PR ^^

**Comment by oli-obk** [comment]
Turns out we didn't even have `simd_shuffle` tests in miri. Added both.

**Comment by RalfJung** [comment]
I'm pretty sure we had such tests in Miri, but probably only via portable-simd APIs. Cannot hurt to have some direct tests as well for special intrinsics like this one. :)

**Comment by workingjubilee** [comment]
Nice!

@bors r+

File: compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs [hunk 3]
```diff
@@ -117,6 +117,54 @@ pub(super) fn codegen_simd_intrinsic_call<'tcx>(
...
+            let (lane_count, lane_ty) = layout.ty.simd_size_and_type(fx.tcx);
+            let (ret_lane_count, ret_lane_ty) = ret.layout().ty.simd_size_and_type(fx.tcx);
+
+            assert_eq!(lane_ty, ret_lane_ty);
+            assert_eq!(idx.len() as u64, ret_lane_count);
+
+            let total_len = lane_count * 2;
+
+            let indexes =
+                idx.iter().map(|idx| idx.unwrap_leaf().try_to_u16().unwrap()).collect::<Vec<u16>>();
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs:151]: This should have used u32. Will fix in https://github.com/bjorn3/rustc_codegen_cranelift

**Pull request merged**